File: backend/routes/members.py
from fastapi import APIRouter, Depends, HTTPException
from repositories.members import member_repo
from schemas.member import MemberCreate, MemberUpdate, MemberResponse
from utils.response import ApiResponse

router = APIRouter(prefix="/members", tags=["Members"])


@router.get("", response_model=ApiResponse)
async def get_members(project_id: int = None):
    """Get all members, optionally filtered by project_id"""
    if project_id:
        members = member_repo.get_by_project(project_id)
    else:
        members = member_repo.get_all()
    return ApiResponse.ok(data=members)


@router.get("/{member_id}", response_model=ApiResponse)
async def get_member(member_id: int):
    """Get a member by ID"""
    member = member_repo.get_by_id(member_id)
    if not member:
        raise HTTPException(status_code=404, detail="Member not found")
    return ApiResponse.ok(data=member)


@router.post("", response_model=ApiResponse)
async def create_member(member_in: MemberCreate):
    """Create a new member"""
    # Check if user is already in the project
    existing = member_repo.get_by_field("user_id", member_in.user_id)
    if existing and existing.get('project_id') == member_in.project_id:
        raise HTTPException(status_code=400, detail="User is already in this project")

    member = member_repo.create(member_in.model_dump())
    return ApiResponse.ok(
        data=member, message="Member created"
    )


@router.put("/{member_id}", response_model=ApiResponse)
async def update_member(
    member_id: int,
    member_in: MemberUpdate,
):
    """Update a member"""
    existing = member_repo.get_by_id(member_id)
    if not existing:
        raise HTTPException(status_code=404, detail="Member not found")
    updated = member_repo.update(member_id, member_in.model_dump(exclude_unset=True))
    return ApiResponse.ok(
        data=updated, message="Member updated"
    )


@router.delete("/{member_id}", response_model=ApiResponse)
async def delete_member(member_id: int):
    """Delete a member"""
    member = member_repo.get_by_id(member_id)
    if not member:
        raise HTTPException(status_code=404, detail="Member not found")
    member_repo.delete(member_id)
    return ApiResponse.ok(message="Member deleted")

File: backend/src/routes/members.ts
import type { FastifyPluginAsync } from 'fastify';

import { BaseRepository } from '../repositories/base.js';
import { successResponse, errorResponse } from '../utils/response.js';

type ParamsWithId = { id: string };
type StoredBaseRecord = Record<string, unknown> & { id: number; created_at: string; updated_at: string };

const memberRepo = new BaseRepository<StoredBaseRecord>('members.json');

function parseNumber(value: string) {
  return Number.parseInt(value, 10);
}

export const memberRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get('/', async (request) => {
    try {
      return successResponse(await memberRepo.findAll());
    } catch (error) {
      request.log.error(error);
      return errorResponse('Failed to get members');
    }
  });

  fastify.get<{ Params: ParamsWithId }>('/:id', async (request, reply) => {
    try {
      const member = await memberRepo.findById(parseNumber(request.params.id));
      if (!member) {
        reply.code(404);
        return errorResponse('Member not found');
      }
      return successResponse(member);
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to get member');
    }
  });

  fastify.post('/', async (request, reply) => {
    try {
      const member = await memberRepo.create(request.body as Record<string, unknown>);
      return successResponse(member, 'Member created');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to create member');
    }
  });

  fastify.put<{ Params: ParamsWithId }>('/:id', async (request, reply) => {
    try {
      const updated = await memberRepo.update(parseNumber(request.params.id), request.body as Record<string, unknown>);
      if (!updated) {
        reply.code(404);
        return errorResponse('Member not found');
      }
      return successResponse(updated, 'Member updated');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to update member');
    }
  });

  fastify.delete<{ Params: ParamsWithId }>('/:id', async (request, reply) => {
    try {
      const deleted = await memberRepo.delete(parseNumber(request.params.id));
      if (!deleted) {
        reply.code(404);
        return errorResponse('Member not found');
      }
      return successResponse(null, 'Member deleted');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to delete member');
    }
  });
};
